qiskit/dagcircuit/_dagcircuit.py

```diff
@@ -881,7 +881,7 @@ def substitute_node_with_dag(self, node, input_dag, wires=None):
         if condition:
             input_dag.add_creg(condition[0])
             to_replay = []
-            for n_it in nx.topological_sort(input_dag.multi_graph):
+            for n_it in input_dag.multi_graph:
```

Comment:
Is there a risk here of re-assembling the input DAG in the wrong order? Does `dag.__iter__` provide any guarantees on order enumeration order?

Unlike the case below, where we first capture `_full_pred_succ_maps` and reconstruct the graph from the stored edges, here we first decompose to an ordered list of ops, and then `apply_operation_back`, so the order of `to_replay` is important.

Reply:
I'm undoing this one in 10ddb4a0 so we can merge this PR and have a deeper look to your concern, that seems very valid.